### ISSUE TITLE
Fix not revealing stored passwords from the UI

### ISSUE THREAD

**Issue by kenkendk** [duplicati/duplicati#2024]
Before we can release the v. 2.0 beta we should fix the current insecure practice with storing passwords in "almost plain" and also prevent revealing the password from the UI.

There are a number of issues relating to this, and I am closing the imported ones:

  - #558 - A master password
  - #703 - Read-only access
  - #1010 - Hide auth password
  - #1251 - A global password
  - #1942 - Prevent exporting passphrase in JSON
  - #1807 - Login password protects configuration

A shallow protection would be to simply mask out any "password" marked options, and deliver them as a placeholder string (e.g. `********`).

Another, more thorough approach, would be to use the system keychain to store the password options.

Perhaps we need both, as we may lack access to a keychain on some system, and maybe also for users with `--portable-mode` enabled.

When storing with the keychain, we need to mark the passwords in the database to know where they are stored in the keychain. If we store the passwords in the database, we should add some kind of encryption to the passphrases.

**Comment by telemaxx** [comment]
**After Installation no Password:**
![pw1](https://cloud.githubusercontent.com/assets/22706487/19696561/ef6027aa-9ae7-11e6-9497-96fffbbb4e97.PNG)
**then i set a global Password:**
![pw2](https://cloud.githubusercontent.com/assets/22706487/19696566/f2d02444-9ae7-11e6-8b41-cebda84217b6.PNG)
**Now i have to login:**
![pw3](https://cloud.githubusercontent.com/assets/22706487/19696568/f5270a64-9ae7-11e6-84a8-cd4a27668f8f.PNG)
**then i look at my settings, no Password i marked. But its active.
i would expect Password i checked.**
![pw4](https://cloud.githubusercontent.com/assets/22706487/19696570/f6fbd6a8-9ae7-11e6-8e36-bb05016ffc86.PNG)
**If i change something in the settings and click save, password is not longer active**

**Comment by kenkendk** [comment]
@telemaxx: That is something different, and a bug in the UI

**Comment by telemaxx** [comment]
OK, i create a new ticket.

**Comment by Pectojin** [comment]
@kenkendk I think this issue #3027 should be added to the list (as well as any other advanced options containing passwords)

**Comment by wjansenw** [comment]
Wouldn't it be an option to use https://www.zetetic.net/sqlcipher/open-source/ for the configuration database? You could use an encryption key only known by the Duplicati team that will never be exposed.  

And provide options to export the db unencrypted. 

Haven't read in detail, and don't know if the code could easily be added to this project.

**Comment by Pectojin** [comment]
@wjansenw there is no way to keep such an encryption key secret since it needs to somehow be used by the Duplicati Server when opening the database. Any such usage would be in the code, for everyone to see, so it would no longer be secret.

The solution has to be secured by a secret only known to the user or local machine. E.g. unlocked by a user when starting Duplicati or using the system keychain.

**Comment by wjansenw** [comment]
Yes, it has to be in the code, but does it have to be public? Build the
binaries with a certain key that is not publicly known. The key is now in
the binaries. Yes, there are probably ways to get the key from the
binaries, but that won't be nearly as easy as it is now to get your hands
on all your cloud credentails.

I believe the link explicitly refers to configuration dbs. So something
must be possible here, no?

Wim

**Comment by Pectojin** [comment]
Well, yes it's possible. 

It just has multiple implications such as trusting the Duplicati team with the knowledge of the key used to encrypt your database. 
And to some extend it's accepting obfuscation in place of proper security. I know it *is* encryption, but encryption with a known key is basically just adding an obfuscation layer since it's then just a matter of extracting the key.

It can be done and it is more secure than nothing, but I think we can solve it in a better way.

**Comment by kenkendk** [comment]
Packing a key inside the binaries adds no security. You can dump all strings from most file types with a standard tool:
http://split-code.com/strings2.html

Encrypting the database adds a number of usability issues, which is fine if there is an upside, but I do not see it.

Since the running program needs the key, there is no where to hide it. Using the keychain to store the sensitive data gives the optimal protection (essentially trusting the OS to do it right). In this case it will only be accessible if the keychain is unlocked or the program is running (memory dump).

**Comment by SystemOverlord** [comment]
I see that the issue is still open but the v2.0 beta was already released, are there any other related issues in which I can help? I am a student employee of University Heidelberg, and I would like to offer my help regarding IT security problems.

**Comment by piegamesde** [comment]
@SystemOverlord There are a bunch of keystores, keyrings and password safes out there. The possibility to use one of them instead of storing them in plain text would be awesome. (See #3093)

Alternatively, the login system seems to be not _that_ secure. This isn't really a problem for local login, but could still be improved.

**Comment by Pectojin** [comment]
@piegamesde I'm afraid it's hard to improve on the login system without adding native SSL to Duplicati. The login system already uses salts and nonces to avoid sending the password and to limit replay attacks as much as possible (given that authentication happens over unencrypted HTTP).

In addition to keystore/keyring there's a couple of loose ends with passwords displayed in frontend and then there's the issue with the DB's being plaintext :)

**Comment by SystemOverlord** [comment]
@Pectojin @piegamesde Thank you very much for the hints. I'll have a look and discuss the next steps with my professor.

**Comment by emorgoch** [comment]
Any update on the UI side of this? Testing Duplicati for my own use now, and I couldn't believe that the password field when using a local folder destination wasn't masked. If not, I'd find it interesting to get my feet wet with GitHub finally, and take a stab at clearing as many of those locations out.

**Comment by Pectojin** [comment]
@emorgoch, I haven't heard anything so by all means feel free to take a look

**Comment by sylerner** [comment]
Is there any keyring technology available that uses a common API for Linux (KDE and Gnome at a minimum), Mac and Windows?

I'd like to put a bounty on this, but have no idea how much is appropriate.  In order to gauge the complexity of the task - and hence what would be an appropriate bounty - it's necessary to know if this can be implemented once or would need 3+ implementations.

**Project leaders**:  If a universal keyring manager isn't available,  what would your feelings be about having duplicati (at least initially) making use of a keyring only on Linux?

Thanks!

**Comment by Pectojin** [comment]
There isn't any universal option, so the best option is to write a common interface in Duplicati and implement it for each operating system / keychain. It's fine for it to only work on one operating system initially if that's what the developer is used to working with.

**Comment by sylerner** [comment]
I'm posting a $50 bounty to have **all** passwords on KDE Linux systems stored in KDE Wallet (or other KDE compatible keyring).

**Comment by fergbrain** [comment]
Duplicacy uses the gnome-keyring (https://github.com/gilbertchen/duplicacy/wiki/Managing-Passwords) which uses a d-bus. This should be easy to interface with.

**Comment by Pectojin** [comment]
There seems to be an implementation of "A global password" that never got upstreamed here: https://github.com/myENA/duplicati/commit/52e31a0429f8e0453ee0ff3211820716e1da6fd0

I haven't had time to test it, but it looks ok.

**Comment by Pectojin** [comment]
Hmm, I tested it now, it's missing all the frontend components. So it doesn't really work unless I disable two front end checks, and then we end up with a less user friendly error from the server 

`Unable to save schedule or backup object: The given key 'passphrase' was not present in the dictionary.`

**Comment by sebastianhaberey** [comment]
Hey guys, thank you for your work on Duplicati! 

I'm posting a $50 bounty to have **all** passwords on macOS stored in the standard system keychain. 
If a cross-platform solution is feasible, it would be the preferred way to go, of course.

**Comment by auwsom** [comment]
My workaround is to use the portable installation mode which moves the AppData folder inside the main program folder, then move that to a VeraCrypt volume. I use Smart Retention, a 3hr setting and an autorun.inf to run Duplicati when the volume opens. It should start missed backups as soon as it starts. Here’s a post with some links at SE: https://softwarerecs.stackexchange.com/a/55070/30586

Not as simple for a user as a builtin keyring, but cross-platform, although it would still be exposed when that volume is open. At least it offers strong protection while the volume is closed.

**Comment by digulla** [comment]
See also #3396 for some ideas how to get the passphrase from a command or other sources.

**Comment by digulla** [comment]
I've just checked the last open issue #1251 (see the first post). It's closed but not ruled out. Why? What is left to do for this issue?

**Comment by Pectojin** [comment]
Looks like I closed that one in error. The provided fix is only partial and not a desired solution.

**Comment by Sogolumbo** [comment]
What do you think about only storing the following on disk:
- master password hash
- backup password encrypted by the master password
- backup password hash (if helpful/needed to allow easier recovery when the master password is lost. This would only be stored so that the right backup password can be typed in and duplicati can check whether it's right).

Then there's two options:
- not storing the master password anywhere, waiting for the user to type the master password in
- getting the master password from the keyring of the OS (or another secure storage).

**Comment by ts678** [comment]
> currently the master password is only used to lock the Web interface

There is nothing claiming to be a master password, as per the first link in the original post. That issue rolled into this one.

There is an `Access to user interface` password in the `Settings` screen, and it does what one expects from its name.
[Duplicati.Server.exe](https://duplicati.readthedocs.io/en/latest/07-other-command-line-utilities/#duplicatiserverexe) and relatives call this the `--webservice-password` but there is also `--server-encryption-key` that
could possibly be scripted to a custom (not default) value which is obtained somehow (such as user or a favorite storage)
when Duplicati is started. This relies on system having an encryption-capable SQLite. Windows gets one. Linux might not.
I'd note that the encryption that an encrypted SQLite has isn't high-quality, but it impedes peeking and string scanning...

If there were a master password, it should protect more than the backup password. Storage credentials are also sensitive.
Encrypting the whole database avoids having to figure out what needs protection. Too bad there's the SQLite limitation...

**Comment by Sogolumbo** [comment]
Thanks for the clarification.

I think it would be a good Idea to clarify this in the application and [here](https://github.com/duplicati/duplicati/wiki/Developer-documentation#encryption).

Consider the following scenario:
For protection against encrypting malware we choose to never let the "main system" (e.g. a desktop computer that we want to backup) touch the backups. Instead we take a USB HDD, on which we install a fresh "backup OS" and on which we will also put the backups. We will only connect the USB HDD after the main system has powered down and boot straight to the backup OS and do the backups from there.
If the backup passwords are easily accessible the encryption of the backup is useless.
If you were a journalist who tries to protect your sources from an oppressive regime the clarification above will make a difference of night and day.

**Comment by duplicatibot** [comment]
This issue has been mentioned on **Duplicati**. There might be relevant details there:

https://forum.duplicati.com/t/are-backup-encryption-passphrases-always-stored-in-plain-text/14433/3

**Comment by duplicatibot** [comment]
This issue has been mentioned on **Duplicati**. There might be relevant details there:

https://forum.duplicati.com/t/kestrel-is-coming/15527/22

**Comment by duplicatibot** [comment]
This issue has been mentioned on **Duplicati**. There might be relevant details there:

https://forum.duplicati.com/t/google-drive-forbidden-403/17356/3
